src/ServiceControl.Transports.AzureServiceBus/QueueLengthProvider.cs

```diff
@@ -0,0 +1,115 @@
+namespace ServiceControl.Transports.AzureServiceBus
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Linq;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.ServiceBus;
+    using Microsoft.ServiceBus.Messaging;
+    using Monitoring;
+    using Monitoring.Infrastructure;
+    using Monitoring.Messaging;
+    using Monitoring.QueueLength;
+    using NServiceBus.Logging;
+    using NServiceBus.Metrics;
+
+    public class QueueLengthProvider : IProvideQueueLength
+    {
+        ConcurrentDictionary<EndpointInstanceId, string> endpointQueueMappings = new ConcurrentDictionary<EndpointInstanceId, string>();
+
+        QueueLengthStore queueLengthStore;
+        NamespaceManager namespaceManager;
+
+        CancellationTokenSource stop = new CancellationTokenSource();
+        Task pooler;
+
+        public void Initialize(string connectionString, QueueLengthStore store)
+        {
+            this.queueLengthStore = store;
+            this.namespaceManager = NamespaceManager.CreateFromConnectionString(connectionString);
+        }
+
+        public void Process(EndpointInstanceId endpointInstanceId, EndpointMetadataReport metadataReport)
+        {
+            endpointQueueMappings.AddOrUpdate(
+                endpointInstanceId,
+                id => metadataReport.LocalAddress,
+                (id, old) => metadataReport.LocalAddress
+            );
+        }
+
+        public void Process(EndpointInstanceId endpointInstanceId, TaggedLongValueOccurrence metricsReport)
+        {
+        }
+
+        public Task Start()
+        {
+            stop = new CancellationTokenSource();
+
+            pooler = Task.Run(async () =>
+            {
+
+                while (!stop.Token.IsCancellationRequested)
+                {
+                    try
+                    {
+                        Logger.DebugFormat("Querying namespace manager: {0}", namespaceManager.Address);
+
+                        var queues = await namespaceManager.GetQueuesAsync();
```

Comment:
I deliberately started with a naive implementation. There is an overload of GetQueuesAsync which accepts a filter but I'm not sure whether we can create a filter like `where queueName in (listOfQueueNames)`. I don't even know if you can combine filters with an OR. If not then we'd have to query for each queue individually and at what point does that get more expensive than just doing a single query for all and filtering locally.

We are using the queue names returned by the `QueueBindings` class in the core. I don't know if that contains the already transformed queue names or if the transport translates them afterwards. I know for SQL we had to transform this queue name based on the version of the transport sending us data.

Reply:
OK, so I checked and we are not using QueueBindings. We are instead using `.LocalAddress()`. That address has been through the transport address sanitization process so it will be the actual physical address.

Reply:
> I'm not sure whether we can create a filter like where queueName in (listOfQueueNames). I don't even know if you can combine filters with an OR.

We can't. It's AND only.

Reply:
AccessedAt will return anything involved in send/received operations on the namespace, including queues that are not NServiceBus 🙂

Reply:
@SeanFeldman Yes but is it likely that the list will be significantly smaller than just calling GetQueues with no filter?

The problem is, we may have to gather queue length from 10s-100s of queues and all we have of them is their name. It seems like we can do a single query to get everything (which may be too much) or we can query for each queue individually (which may turn in to a lot of individual queries).

I believe that it would be safer to do the former (single query that gets too much data). Do you agree with that assessment or should we still be looking for a way to reduce that query payload?

Reply:
Agree. Still wish we could define a list of queues to query.

There's also another wrinkle to this - with ASB transport running on EndpointOriented topology, subscriptoins hold messages similar to input queues. Meaning that messages are spread among queues and subscriptoins. In that case, reporting an endpoints' queue length would need to take into consideration subscriptions length as well. Which becomes a bit messy. How do we tackle that one?

Reply:
Does the endpoint read directly from the subscriptions or do the subscription messages end up in the input queue before the endpoint gets them. If the latter then I say we ignore them for now.